pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/AbstractNode.java

```diff
@@ -418,4 +419,36 @@ public GenericToken jjtGetLastToken() {
     public void jjtSetLastToken(GenericToken token) {
         this.lastToken = token;
     }
+
+    @Override
+    public void remove() {
+        // Detach current node of all its children, if any
```

Comment:
why do we need to do this? can't I just remove a subtree and add it as is somewhere else? I believe this will actually make operations harder.

For instance, I want to replace a Node instance but keep it's children untouched. If I remove first, I loose the children. If I remove last, the children will be detached of it's parent (whichever it may be). I'd have to remove the child nodes, one by one, and add them to the new parent, but that would translate the problem to grandchildren...

Suppose I want to reorder nodes (ie, to autofix a rule regarding field / constructor / method declaration ordering). The easiest thing would be to take the out of place node (`ASTFieldDeclaration`, `ASTMethodDeclaration`, etc.) and put it back in at the correct place. This method once again would make that impossible to do. I'd have to add an extra method to "swap" elements, meaning more boilerplate and harder implementations for other grammar engines such as Antlr4 (soon to come to PMD /cc @matifraga @tomidelucca )

Reply:
@jsotuyod I got your point.

This has been done so because it has been the less intrusive way that I have found to stop the visitor pattern through the recently removed tree, since all language-specific abstract nodes seem to have the visitor chain implemented as follows: if the current node's children array is not null it is iterated so as to visit all its children.

The solution I have come up with to handle both cases (supporting the usage you mention and stopping the visitor pattern) is to not detach the children of the node being removed and to add a new field to the abstract node class, a kind of `areChildrenAccepted`, with its default value being true, so as to update all language-specific abstract nodes to check this condition too before visiting node's children. 
Then, when a node is removed, the only thing we have to do to stop the visitor is to set this field to false. When a node is add as a child of another node, we should set this field in true. Caution should be taken if a root node is removed and it wants to be enabled again as root node, as the `areChildrenAccepted` field should be explicitly set to true - however, I think this case is much less common than the other ones mentioned.

What do you think about this?

Reply:
Ok, I get a better picture now.

The bottom line here however is: when do you intend to apply the change to the AST? Altering the AST during visitation has deeper problems than what happens to the children, but also, what happens to parents / siblings.

As the code stands, suppose I have a root node with 2 children. Upon visiting the first one, this one will remove itself from the tree.

This will happen as follows:

1. we visit the root node
2. we call [`childrenAccept`](https://github.com/pmd/pmd/blob/master/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/AbstractJavaNode.java#L53-L59) on the root node
3. we [visit the first node](https://github.com/pmd/pmd/blob/master/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/AbstractJavaNode.java#L56), call remove, and the parent is updated to have only 1 children
4. as the `childrenAccept` loop resumes, our current index (`i`) is incremented from 0 to 1, we check against `children.length` (now 1 since we removed a node), and stop.

We never visited the second child.

I think it would be far saner to never attempt a tree modification during transversal. The autofix needs to be in 2 stages:

1. we visit all nodes and detect violations.
2. we apply fixes.
3. we move on to the next rule

Notice that `ParametricRuleViolation`, which is the base class for `RuleViolations` does receive both the `Rule` and `Node` as parameters, allowing you to eventually come up with autofixes later on. Just *never* keep the `Node` indefinitely, since the `RuleViolation` instance is kept during the whole PMD execution and that would prevent all ASTs from being garbage collected.

Reply:
Good catch & idea! 

I'll leave a reference to this comment in the main issue.
I'll then remove the detachment of all the children.

As the `remove` method may be called during transversal (although it shouldn't), what do you think about adding some kind of `lock` for node modifications during transversal in another PR? 
This lock is also intended to work with the `add` and `replace` methods that we are going to develop.

Just let me know and I'll add that feature to the main issue too :smile:

Reply:
I thought about the lock too. I like the idea in general, I'm hesitant as to how to effectively implement it...

- adding a boolean field to all nodes implies a big memory footprint
- having it at the root of the AST means navigating up the tree all the time to check it

I think it needs some thinking, but definitely not part of this PR

Reply:
Great! Then I'll add that feature to the main issue. 
We may discuss the implementation details soon. I'll let you know if I come up with any idea on how to solve this.